net/tcp/tcp_send_buffered.c

```diff
@@ -604,170 +603,116 @@ static uint16_t psock_send_eventhandler(FAR struct net_driver_s *dev,
+      if (++TCP_WBNRTX(wrb) >= TCP_MAXRTX)
+        {
+          nwarn("WARNING: Expiring wrb=%p nrtx=%u\n",
+                wrb, TCP_WBNRTX(wrb));
 
-              /* The maximum retry count as been exhausted. Remove the write
-               * buffer at the head of the queue.
-               */
+          /* The maximum retry count has been exhausted. Remove the write
+           * buffer at the head of the queue.
+           */
 
+          if (rexmit_from_write_q)
+            {
+              FAR struct tcp_wrbuffer_s *tmp;
               tmp = (FAR struct tcp_wrbuffer_s *)sq_remfirst(&conn->write_q);
               DEBUGASSERT(tmp == wrb);
               UNUSED(tmp);
+            }
 
-              /* And return the write buffer to the free list */
+          /* And return the write buffer to the free list */
 
-              tcp_wrbuffer_release(wrb);
+          tcp_wrbuffer_release(wrb);
 
-              /* Notify any waiters if the write buffers have been
-               * drained.
-               */
-
-              psock_writebuffer_notify(conn);
+          /* Notify any waiters if the write buffers have been
+           * drained.
+           */
 
-              /* NOTE expired is different from un-ACKed, it is designed to
-               * represent the number of segments that have been sent,
-               * retransmitted, and un-ACKed, if expired is not zero, the
-               * connection will be closed.
-               *
-               * field expired can only be updated at TCP_ESTABLISHED state
-               */
+          psock_writebuffer_notify(conn);
 
-              conn->expired++;
-            }
-        }
-
-      /* Move all segments that have been sent but not ACKed to the write
-       * queue again note, the un-ACKed segments are put at the head of the
-       * write_q so they can be resent as soon as possible.
-       */
+          /* NOTE expired is different from un-ACKed, it is designed to
+           * represent the number of segments that have been sent,
+           * retransmitted, and un-ACKed, if expired is not zero, the
+           * connection will be closed.
+           *
+           * conn->expired can only be updated in TCP_ESTABLISHED state.
+           */
 
-      while ((entry = sq_remlast(&conn->unacked_q)) != NULL)
-        {
-          wrb = (FAR struct tcp_wrbuffer_s *)entry;
-          uint16_t sent;
+          conn->expired++;
 
-          /* Reset the number of bytes sent sent from the write buffer */
+          /* Continue waiting */
 
-          sent = TCP_WBSENT(wrb);
-          if (conn->tx_unacked > sent)
-            {
-              conn->tx_unacked -= sent;
-            }
-          else
-            {
-              conn->tx_unacked = 0;
-            }
-
-          if (conn->sent > sent)
-            {
-              conn->sent -= sent;
-            }
-          else
-            {
-              conn->sent = 0;
-            }
+          return flags;
```

Comment:
remove

Reply:
Could you please explain why it should be removed here?

Reply:
emm... I am comment on the wrong line, remove line 715:
https://github.com/apache/incubator-nuttx/pull/5589/files/02c2b602c055214b111969e5ab207b77325f77c5#diff-ac1dd416b803b7f5b3118a05af87bdcef81e8b77110411711ec801bac928d274R715

flags will return on line 732:
https://github.com/apache/incubator-nuttx/pull/5589/files/02c2b602c055214b111969e5ab207b77325f77c5#diff-ac1dd416b803b7f5b3118a05af87bdcef81e8b77110411711ec801bac928d274R732

Reply:
> flags will return on line 732:

Then case tcp_sendbuffer_notify(conn) will be invoked. Why its invocation is necessary if we are retransmitting a segment?

Reply:
It depends on whether tcp_wrbuffer_release is called before tcp_sendbuffer_notify, it seems that you return flags early after release, which is different from the original code logic

Reply:
Compared to the old code, the new code does not modify anything in write_q. Thus tcp_sendbuffer_notify(conn) invocation is not reasonable anymore if we are retransmitting a segment.

Reply:
Yes, you released the wrbuffer here without tcp_sendbuffer_notify on line 659

https://github.com/apache/incubator-nuttx/pull/5589/files/02c2b602c055214b111969e5ab207b77325f77c5#diff-ac1dd416b803b7f5b3118a05af87bdcef81e8b77110411711ec801bac928d274R659